Add comprehensive test suite for Feed-Forward Neural Network wavefunction (David’s implementation)

q-pratz-chem opened the pull request:
This PR introduces a new test file covering the Feed-Forward Neural Network (FFNN) wavefunction implemented by David.

- The test suite achieves **100% code coverage** for this module.
- The addition is primarily **for bookkeeping and archival purposes**—the implementation is retained in the main branch for reference.
- **Note:** The physical or numerical performance of this wavefunction has **not yet been validated**. If future work aims to verify or improve the model’s functionality, such efforts should be conducted in a **separate development branch**.

q-pratz-chem committed:
Implement tests for RBM wfn by David written in feed-forward neural network form

q-pratz-chem committed:
Update tests for better coverage

q-pratz-chem committed:
Update tests for feed-forward-RBM to ensure 100% test coverage

kzsigmond commented on tests/test_wfn_network_rbm_feed-forward-NN.py:
```diff
@@ -0,0 +1,264 @@
+
+    sds = [0b0011, 0b1100]
+    # pick a parameter index to compare
+    deriv_index = 5 if rbm.nparams > 6 else 0
```
which case are we testing here? `rbm.nparams` should be fixed, so we are either using `deriv_index` 5 or 0.

q-pratz-chem replied:
I have corrected it, we are using 0.

kzsigmond commented on tests/test_wfn_network_rbm_feed-forward-NN.py:
```diff
@@ -0,0 +1,264 @@
+    rbm.assign_params(plus); f_plus = rbm.get_overlap(sd0)
+    rbm.assign_params(minus); f_minus = rbm.get_overlap(sd0)
+    numeric = (f_plus - f_minus) / (2 * eps)
+    rbm.assign_params(base_flat)
```
could we use the function `finite_diff_gradient` here?

kzsigmond reviewed (commented):
Thanks for writing the tests for RBM! Could you please go over the comments?

q-pratz-chem commented on tests/test_wfn_network_rbm_feed-forward-NN.py:
```diff
@@ -0,0 +1,264 @@
+        assert p.shape == shape
+
+
+def test_assign_params_flat_and_influences_overlap():
+    rbm = RestrictedBoltzmannMachine(nelec=2, nspin=4, nbath=2, orders=(1,))
+    flat = np.zeros(rbm.nparams); flat[0] = 0.2
+    rbm.assign_params(flat)
+    olp0 = rbm.get_overlap(0b0)
+    olp1 = rbm.get_overlap(0b1)
+    assert olp0 == pytest.approx(1.0)
+    assert not np.isclose(olp1, 1.0)
```
The test is correct in its place.
 
The overlap is one in the case of zero electrons because `overlap = Π_j σ( w_j · m ) * output_scale`, where `σ(x) = 1 + exp(x)` and `output_scale = 0.5`.

If the correct physical behavior is expected, then there is a need to enforce:

- `get_overlap(sd)=0 when popcount(sd) != nelec`
- or incorporate antisymmetry.

kzsigmond replied:
Raised issue #185 so that we do not lose track of this.

q-pratz-chem commented on tests/test_wfn_network_rbm_feed-forward-NN.py:
```diff
@@ -0,0 +1,264 @@
+    assert not np.isclose(olp1, 1.0)
+
+
+def test_assign_params_invalid_length():
+    rbm = RestrictedBoltzmannMachine(nelec=2, nspin=4, nbath=2)
+    original = rbm.params.copy()
+    longer = np.zeros(rbm.nparams + 1)
+    rbm.assign_params(longer)  # should not crash
+    # Parameters should still have expected shape and size
+    assert rbm.params.shape == original.shape
+    assert np.isfinite(rbm.params).all()
```
There is no conceptual reason this wavefunction should accept parameter arrays that are longer than required. The current behavior, silently truncating extra entries, is simply a consequence of how David originally wrote the slicing logic inside `assign_params`.

If stricter behavior is preferred, an explicit length check is required in place, so that assigning too-long arrays raises a ValueError instead of being silently truncated.

kzsigmond replied:
Raised issue #185 so that we do not lose track of this.

q-pratz-chem committed:
Edited tests as per the PR review

q-pratz-chem commented on tests/test_wfn_network_rbm_feed-forward-NN.py:
```diff
@@ -0,0 +1,264 @@
+    assert overlaps.shape == (len(sds),)
+
+    # Derivatives w.r.t multiple params
+    for deriv_index in [0, rbm.nparams - 1]:
```
I have updated the test accordingly.

kzsigmond committed:
chore: remove unused slater import

kzsigmond committed:
chore: remove hard coded deriv index

Deriv index is defined above, but only used for one of the derivatives, while it is hard coded for the other.

kzsigmond committed:
chore: uncomment derivative check w.r.t multiple parameters

kzsigmond commented on tests/test_wfn_network_rbm_feed-forward-NN.py:
```diff
@@ -0,0 +1,264 @@
+    assert np.isfinite(norm_after) and norm_after > 0
+
+
+@pytest.mark.xfail(reason="pspace=None not supported yet")
+def test_normalize_none():
+    rbm = RestrictedBoltzmannMachine(nelec=2, nspin=4, nbath=2)
+    rbm.normalize(pspace=None)
```
Raised issue #185 so that we do not lose track of this.

kzsigmond reviewed (approved):
@q-pratz-chem I made some minor changes and created an issue to some up some issues that belong to the RBM, not the unit test for the RBM. If everything looks good, you can go ahead and merge it.
